Sources/DarkModeCore/NSObject+DarkModeKit.m

```diff
@@ -7,23 +7,51 @@
 
 @import ObjectiveC;
 
-@implementation NSObject (DarkModeKit)
+@interface IMPContainer ()
 
-+ (BOOL)dm_swizzleInstanceMethod:(SEL)selector1 to:(SEL)selector2 {
-  Method method1 = class_getInstanceMethod(self, selector1);
-  Method method2 = class_getInstanceMethod(self, selector2);
+@property (readwrite) IMP imp; // Redefine as readwrite so we can set it in this file
 
-  if (!method1 || !method2) {
-    return NO;
-  }
+@end
+
+@implementation IMPContainer
 
-  if (class_addMethod(self, selector1, method_getImplementation(method2), method_getTypeEncoding(method2))) {
-    class_replaceMethod(self, selector2, method_getImplementation(method1), method_getTypeEncoding(method1));
-  } else {
-    method_exchangeImplementations(method1, method2);
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _imp = NULL;
   }
+  return self;
+}
+
+@end
+
+OBJC_EXPORT id objc_msgSendSuper2(struct objc_super *super, SEL op, ...);
+
+@implementation NSObject (DarkModeKit)
+
++ (BOOL)dm_swizzleSelector:(SEL)selector withBlock:(id (^)(IMPContainer *))implementationCreationBlock {
+  IMPContainer *container = [IMPContainer new];
+  IMP newIMP = imp_implementationWithBlock(implementationCreationBlock(container));
+  // Wrap original implementation in an object so it
+  // can be retrieved in the swizzled implementation
+  container.imp = swizzleSelector(self, selector, newIMP);
+  return container.imp != NULL;
+}
+
+static IMP swizzleSelector(Class clazz, SEL selector, IMP newImplementation) {
+  Method method = class_getInstanceMethod(clazz, selector);
+  if (!method)
+    return NULL;
+
+  const char *types = method_getTypeEncoding(method);
+  // Make sure the class implements the method. If this is not the case, inject an implementation, only calling 'super'.
```

Comment:
I think we shouldn't implement a **very** generic swizzling method in this project. Implementation details like `objc_msgSendSuper2` makes me feel the maintenance complexity increased.

I'd suggest we using tests or checking if implementation exists via runtime methods to prevent the "no implementation issue".

Reply:
Instead of relying on class_replaceMethod to return original IMP, let's just get method_getImplementation to return the original IMP so as to avoid adding method that calls super.